### ISSUE TITLE
HD path as selection path in sonos like it is with upnp servers?

### ISSUE THREAD

**Issue by Viper525** [henkelis/sonospy#74]
Hello.

most Upnp Server show the option to go through the path on the hdd to get to the data.
Can anybody help me how to configure the indes files for this option?

Thanks
 Viper

**Comment by henkelis** [comment]
Hi,

I'm not sure what you mean.

When you scan your music you create a database of tracks. The music can 
be local or on a network drive. The database can be local or on a 
network drive too. When you run Sonospy you tell it where the database 
is (local or network) - Sonospy then locally serves music metadata from 
the database, redirecting to the music as it is rendered.

Mark.

**Comment by Viper525** [comment]
Hi Mark.

1st:
I have 4 questions concerning the indexing.
I've found three .ini files:
- defaultindex.ini
- userindex.ini
  -userindex2.ini
  Looking into the files and at the sonos windows constroller I'm thinking about if the user*.ini files are relevant or if they are just examples and the programm doesn't look at them. Or: I just don't get behind the strukture. :-/

2nd (Issue Topic):
I want to have an index matching to the directory-path the tracks are stored in the NAS (the same as if you select "Folders" inside of the "music Library") Most Upnp Servers show these Index, too. So I thought Sonospy would be able to do that.

3rd:
I've cahnged the defauldindex.ini (as I didn't understand the structure with the userindex-files). It includex the following index:

tree=artist/track
    title=Artists
    icon=icon_artists
        index=artist
        icon=icon_artists
            index=track
            icon=icon_tracks

When I have this inndex active no index can be loaded. When I comment this index out all other indexes are shown correctly in the controller. So I think there must be a Syntax error, but I really don't get it. .-(

4th:
Is it possible to set different indexes for different "instances" of sonospy? (SMAPI-Server A only shows index A and SMAPI-Server B only shows index b?)

Thanks a lot! It's really a great software, I just hav to get behind some special things. So I really appreciate you taking the time to answer. ;-)

Viper

**Comment by henkelis** [comment]
There's a google group for Sonospy 
(https://groups.google.com/forum/#!forum/sonospy-devel)- some of the 
documentation is a bit out of date, but the latest code changes are 
discussed on the group. They guys on there have some good indexing 
setups too. I'll give you brief answers below.

> 1st:
> I have 4 questions concerning the indexing.
> I've found three .ini files:
> - defaultindex.ini
> - userindex.ini -userindex2.ini Looking into the files and at the
>   sonos windows constroller I'm thinking about if the user*.ini
>   files are relevant or if they are just examples and the programm
>   doesn't look at them. Or: I just don't get behind the strukture. :-/

If you use the default pycpoint.ini file, then Sonospy will use the 
defaultindex.ini file - which is just an externalisation of the default 
rules to create the same indexes as Sonos does. You can change that 
file, but it doesn't offer the flexibility of user indexing.

In pycpoint.ini there are 2 flags you can set, to tell Sonospy to use 
alternate indexing. These are in the [indexing] section, with a full 
description there. alternate_indexing allows you to add extra data into 
the displayed element, or change the sort order (for this you would edit 
defaultindex.ini). user_indexes allows you to create your own indexes 
(for this you create something like userindex.ini, or multiple versions 
if you like). If you use userindex.ini, then that will automatically be 
picked up, if you use a different name then you specify the name on the 
command line after the (each) database name:

```
 -sSonospy=Viper,v.db,viperindex.ini
```

> 2nd (Issue Topic):
> I want to have an index matching to the directory-path the tracks are 
> stored in the NAS (the same as if you select "Folders" inside of the 
> "music Library") Most Upnp Servers show these Index, too. So I thought 
> Sonospy would be able to do that.

Sonospy doesn't currently support the folders view - I did think about 
adding that but nobody else has asked for it :o)

> 3rd:
> I've cahnged the defauldindex.ini (as I didn't understand the 
> structure with the userindex-files). It includex the following index:
> 
> tree=artist/track
> title=Artists
> icon=icon_artists
> index=artist
> icon=icon_artists
> index=track
> icon=icon_tracks
> 
> When I have this inndex active no index can be loaded. When I comment 
> this index out all other indexes are shown correctly in the 
> controller. So I think there must be a Syntax error, but I really 
> don't get it. .-(

I'll have to check why that would not work, though I suspect it's 
because artists have albums, which then have tracks. You should be able 
to do that with a user defined index (using userindex.ini as above) - 
defaultindex.ini is really there so you can add metadata to the indexes 
it gives you, not to change the index structure.

> 4th:
> Is it possible to set different indexes for different "instances" of 
> sonospy? (SMAPI-Server A only shows index A and SMAPI-Server B only 
> shows index b?)

Yes, you can had a separate user index file for each instance - you 
could also use the servicename key to make an index specific to a SMAPI 
server. Within each file, you can also specify different behaviour for 
different controllers.

It's worth reading the group to get your head around user indexes, 
they're quite powerful. There's a fair bit of documentation in the 
userindex.ini file too.

**Comment by Viper525** [comment]
Hi Mark.

that sounds awesome! I'll try to dive into it.

When I'll be able to realize differen indexes for the servers and understand how these works, I think I'll be able to realize what I need with the indexes.

Thank a lot (again) ;-)

**Comment by Viper525** [comment]
It's really strage:
Everytime I activate

tree=artist/track
    title=Artists

The service is not able show the indexes
:-/

**Comment by henkelis** [comment]
That's a bug. I've fixed it and pushed it to msrefactor5.

Mark.

**Comment by Viper525** [comment]
It works! Great abd Thanks a lot!
But now I have a porblem with searches. It always says: "There was a problem searching. Please try again."

But shouldn't this work?

[indexes]

tree=album/track
    title=Albums
    display=N

tree=artist/track
    title=Artists
    display=N

tree=track
    title=Tracks
    display=N

[searches]

search=Albums
    title=Album

search=Artists
    title=Artist

search=Tracks
    title=Track

multisearch=Albums,Artists,Tracks
    title=Album/Albumartist/Track

**Comment by henkelis** [comment]
Yes that should work. Which ini file are those in, what pycpoint.ini 
settings do you have for alternative_indexing and user_indexes, and what 
is your command line?

**Comment by Viper525** [comment]
hm. that's strange now. I realized, that on one pc the searche is still like this:
![03](https://cloud.githubusercontent.com/assets/8208705/3683963/d88fc132-12f1-11e4-94a2-942531480a2e.PNG)

while on another pc it's like that (correct):
![04](https://cloud.githubusercontent.com/assets/8208705/3683968/fea86b80-12f1-11e4-9660-942454bf8752.PNG)

And yes, it's the same service!? :-p

my command line is the following:
./sonospy_proxy -r -sSonospy=MusicBay-Alben,Alben.db,Alben.ini -sSonospy=MusicBay-Interpreten,Interpreten.db,Interpreten.ini -sSonospy=MusicBay-Charts,Charts.db,ChartsSampler.ini -sSonospy=MusicBay-Medleys,Medleys.db,Medleys.ini -sSonospy=MusicBay-Sampler,SamplerSoundtracks.db,ChartsSampler.ini

I've tried the service "MusicBay-Alben", but its the same with the other services.
At least looking at the search indexes.
Another issues seems to be, that not all files are shown. Are there special requests for the meta-data? (perhaps only v2 is read, not v1, or else?)

Here are my ini files in a rar file, renamed to jpg. ;-)
![inis](https://cloud.githubusercontent.com/assets/8208705/3684002/1054b4c8-12f3-11e4-98a4-daf8c29cc621.jpg)
